Add base models

shmashing opened the pull request.

shmashing committed:
Add in Base Models

This commit adds in the base models that will be used for the web app

shmashing committed:
Add Average Eff To Guardian

This commit adds the average efficiency field to the guardian object because i think this will be a field we will want to track over time

shmashing committed:
Averages Usually Arent Ints

This commit changes the average efficiency field from an int to double because why would that ever be an int

shmashing commented:
@jaredonline #letsdothis

jaredonline commented on Destiny.ScrimTracker.Logic/Destiny.ScrimTracker.Logic.csproj:
```diff
@@ -6,7 +6,6 @@
 
   <ItemGroup>
     <Folder Include="Adapters" />
-    <Folder Include="Models" />
```
Can you explain this to me? I'm surprised that when we add the `Models` folder, we delete this line.

shmashing replied:
This was a woopsie on my part

jaredonline commented on Destiny.ScrimTracker.Logic/Models/Guardian.cs:
```diff
@@ -0,0 +1,9 @@
+    {
+        public string GamerTag { get; set; }
+        public int Elo { get; set; }
+        public double AverageEfficiency { get; set; }
```
I don't think we should track these in the Guardian model. I think we should have a `GuardianELO` model, which I'll talk about more in a different text box.

Same with `GuardianEfficiency`

jaredonline commented on Destiny.ScrimTracker.Logic/Models/GuardianMatchResult.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public class GuardianMatchResult
+    {
+        public Guardian Guardian { get; set; }
```
```suggestion
        public Guardian Guardian { get; set; }
        public Match Match { get; get; }
```

jaredonline commented on Destiny.ScrimTracker.Logic/Models/Match.cs:
```diff
@@ -0,0 +1,13 @@
+        public IEnumerable<GuardianMatchResult> AlphaTeam { get; set; }
+        public IEnumerable<GuardianMatchResult> BravoTeam { get; set; }
+        public int AlphaTeamScore { get; set; }
+        public int BravoTeamScore { get; set; }
```
We should track the match type as well.

jaredonline replied:
Also how will this handle the Rumble type where there are no teams?

jaredonline replied:
Thinking futher, I think we should introduce a new model `MatchTeam`, and this model should have a list of `MatchTeam` objects associated with it. This way we can handle many configurations of teams.

shmashing replied:
hmm. i think this makes sense. this would cover the rumble situation. you could then also just loop through, or maybe even have a `MatchTeam Winner` property on the Match

jaredonline reviewed (changes requested):
I don't think we should put any aggregate of stats on the `Guardian` model directly. I think that'll make us lose data, and if we want to change the way we calculate data, we won't have the raw figures to work with.

Instead I think we should have a `GuardianELO` which is a point-in-time representation of the guardian's elo. It could have a `Match` object linked to it that represents the match that corresponds to the new ELO. This way we could chart ELO over time for guardians. 

Same with efficiency, we could chart average over time.

If we want to rollup these stats, we could have a `GuardianRollupStats` object that has a lot of this stuff precalculated for easy access.

Choo think?

shmashing commented:
You can't thread on comments?? da heck??

I think I agree with most of that. I wasn't intending these to be the backend DB relationship models, per-say, I was more intending these to be the frontend objects to display/take in the information.

That said, I do think that we could use these models along with incorporating bitemporality into the DB to be able to track efficiency/ELO as it progresses over time. It will add more rows to the table, but if we're only storing 50 per guardian anyway, I don't think it'll be _that_ much of an issue.

jaredonline commented:
> bitemporality

I'm not sure what this means in this context. Can you explain some more?

shmashing commented:
having the database be the living record for a guadian. for example, when a guardian is created a row like:
`shmashing | 1200 | 1.22 | <record_start_timestamp> | <null>`

then after a match or two, that row is invalidated by:
`shmashing | 1200 | 1.22 | <record_start_timestamp> | <record_stop_timestamp>`

and a new row is added to reflect the new state:
`shmashing | 1300 | 1.00 | <record_start_timestamp> | <null>`

jaredonline commented:
Got ya. What's the advantage of doing it that way over having a `GuardianELO` record? 

> It will add more rows to the table, but if we're only storing 50 per guardian anyway, I don't think it'll be that much of an issue.

I know that we're looking at only taking 50 matches into consideration as a business level concern, but I think we should store results from all guardians for all time. There's no reason to delete them... Postgres can effeciently handle millions of rows as long as we keep our indexes right, and if we ever get the point that it can't, we can shard it and we'd have a massive company on our hands.

shmashing commented:
I dunno! Fewer models to manage. Fewer tables to manage. Easily query a single table to get all of the history for a single guardian

Yea that's fair. That'd be doooope. we'd be rich probably

jaredonline commented:
> Easily query a single table to get all of the history for a single guardian 

That would be possible with a `GuardianELO` table too, I think. `select * from guardian_elos where guardian_id = ?` sort of thing

There's a few reasons I prefer keeping them separate:

- clear separation of concerns; each table is tracking a limited set of isolated information. ELO, effeciency, some other stat, guardian information, etc. Imagine when a guardian changes names (if that can happen?) or some other part of meta-data, you'd have a record for every match that would need to change as well.
- indexability; if each table has a more separate set of concerns, building specific indexes for specific queries will be easier. For example, we can index on guardian name for a search function much easier if they're separate tables (and querying for a single guardian will be yield easier to handle results)
- lack of sparse / redundant tables; a table that has the same information repeated a bunch creates a lot of wasted space, and if there's a column added (say we start tracking a new stat), it will be null for a large portion of matches. Sparse tables also add to wasted space.
- ease of migration; in general the less information you have in a single table the easier it is to manage that table (make changes or do data transformations on)

shmashing commented:
So then you're thinking of having a GuardianELO table and a GuardianEfficiency table? what would those look like?

jaredonline commented:
```
GuardianELO {
  int MatchId
  int GuardianId
  int StartingElo
  int EndingElo
}

GuardianEffeciency {
  int MatchId
  int GuardianId
  float StartingEffeciency 
  float EndingEffeciency
}
```

I'd put appropriate timestamps on both the models and see how far that gets us

shmashing committed:
Break Out Model Concerns

This commit breaks out different guardian conerns from the guardian class.

A guardians elo will now be represented by their own table so we can better track elo changes over times.
A guardians efficiency will also now be tracked by its own table for the same reasons.

This commit also changes the structure of the match object so it is not limited by two teams.

shmashing committed:
Add Models Folder Back to csproj

jaredonline commented on Destiny.ScrimTracker.Logic/Models/Guardian.cs:
```diff
@@ -0,0 +1,9 @@
+    public class Guardian
+    {
+        public string GamerTag { get; set; }
+        public int LifetimeKills { get; set; }
+        public int LifetimeDeaths { get; set; }
```
Should we track these over time in a similar manner to other stats?

shmashing replied:
I figured since they will never be an up and down kinda thing, only an ever increasing number itd be fine to just track the sum. Choo thank?

jaredonline replied:
Seems fine with me unless we want to track it over time for folks?

jaredonline commented on Destiny.ScrimTracker.Logic/Models/GuardianElo.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class GuardianElo
+    {
+        public int MatchId { get; set; }
```
Do you think this should link to the `Match` or `MatchResult` or both? I'm honestly not sure what the best approach here is...

shmashing replied:
I think linking to the match might be better? Since you can get to individual match result from there that seems to be enough information

jaredonline commented on Destiny.ScrimTracker.Logic/Models/MatchType.cs:
```diff
@@ -0,0 +1,11 @@
+namespace Destiny.ScrimTracker.Logic.Models
+{
+    public enum MatchType
+    {
+        Rumble = 1,
+        Control = 2,
+        Clash = 3,
+        Survival = 4,
+        Elimination = 5
+    }
+}
```
😍

jaredonline commented on Destiny.ScrimTracker.Logic/Models/MatchTeam.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public class MatchTeam
+    {
+        public IEnumerable<GuardianMatchResult> Guardians { get; set; }
```
This feels awkward to me... but I'm not sure why. My first thought was "this should be linking to guardians, not the results! But I don't think that's right either...

If you have a match, to get a list of guardians that participated you'd need to go:

`Match -> Match Team -> Match Result -> Guardian` which is a lot of traversal. Maybe that's okay? What do you think?

shmashing replied:
That is a lot of traversal, but i wasn't sure how else to better track all this information. It doesn't seem like _too_ much of a bad thing to me? We could also add a match ID to match result to maintain the two way relationship?

jaredonline replied:
Yea let's def add the ID to the `MatchResult`

jaredonline reviewed (commented):
This is getting a lot closer to something that feels right to me. Nice work!

jaredonline commented:
Oh, a few global comments:

- Every model should have a `GUID`, Globally Unique Identifier. Stripe does a thing that I really like for ID generation we can chat about that guarantees uniqueness across models, so that if all you have is an ID you can still figure out what model you're referencing
- Some models have like `MatchId` which is an `int` and some models have direct reference to another model, like `Guardian` which is of `Guardian` type. Can we standardize one way or the other?

shmashing commented:
Ah good call. I forgot to add Ids to all the objects. We want them to be guids? I'm okay with this. Zapp tried to do something similar with its Ids but failed miserably with it 😂 

Re: objects vs Ids, I feel like it should be based off the hierarchy of the schema. Like `Match` appears to be the highest in the hierarchy for us, so it can have a `MatchTeam` object as a property, but since `MatchTeam` is a child, it needs to have the `MatchId` on it. That make sense? That said, that logic isn't applied everywhere and it should be. i.e. `GuardianMatchResult` should have a `MatchTeamId` and probably a `MatchId` field as well. I think keeping many to many relationship for `Guardian` and `GuardianMatchResult` should be okay. Thoughts?

jaredonline commented:
How did Zapp fail with GUIDs?

shmashing committed:
Add In Base CRUD For Guardians

This commit adds in the base file structure and code to be able to get all guardians, get a specific guardian, create a guardian, and also delete them.

This commit also adds in the logic to update a guardians total kills and total deaths

shmashing commented:
@jaredonline  Kiiiiiinda blew scope here for the branch, but this is what the functionality looks like in C# for all the crud for an object. Lemme know what you think about it. This is the structure conforms to the microsoft standards and patterns (which totally may be overkill for this project).

shmashing commented on Destiny.ScrimTracker.DatabaseMigrator/CommandLineArgs/MigrateOptions.cs:
```diff
@@ -0,0 +1,22 @@
+using CommandLine;
```
ah this should be deleted.

shmashing commented on Destiny.ScrimTracker.DatabaseMigrator/Destiny.ScrimTracker.DatabaseMigrator.csproj:
```diff
@@ -0,0 +1,14 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
This file should also get deleted.

shmashing commented on Destiny.ScrimTracker.DatabaseMigrator/Program.cs:
```diff
@@ -0,0 +1,71 @@
+﻿using System;
```
This file should also get deleted.

shmashing committed:
Remove Old Migration Logic

This commit removes leftover code from how I was thinking about handling migrations before.

Now we will just use entity framework to handle all that jazz

shmashing committed:
Add Get Single Guardian Endpoint

This commit adds in the logix to get a single guardian from the DB.

jaredonline commented on Destiny.ScrimTracker.Api/Controllers/GuardianController.cs:
```diff
@@ -0,0 +1,59 @@
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+
+namespace Destiny.ScrimTracker.Api.Controllers
```
Is there auth on these controllers?

shmashing replied:
Noooo not yet

jaredonline commented on Destiny.ScrimTracker.Api/Controllers/GuardianController.cs:
```diff
@@ -0,0 +1,59 @@
+        private readonly IGuardianService _guardianService;
+        private readonly ILogger<GuardianController> _logger;
+
+        public GuardianController(IGuardianService guardianService, ILogger<GuardianController> logger)
```
Where is this instantiated and how is it instantiated?

shmashing replied:
It's created in Startup.cs using Dependency Injection.

jaredonline commented on Destiny.ScrimTracker.Api/Destiny.ScrimTracker.Api.csproj:
```diff
@@ -1,10 +1,11 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Destiny.ScrimTracker.Logic\Destiny.ScrimTracker.Logic.csproj" />
```
What does this do?

shmashing replied:
Adds a reference to another Project (.csproj file) so we can reference the objects/classes/methods within it.

jaredonline commented on Destiny.ScrimTracker.Api/Migrations/20200103042953_InitialCreate.Designer.cs:
```diff
@@ -0,0 +1,45 @@
+            modelBuilder.Entity("Destiny.ScrimTracker.Logic.Models.Guardian", b =>
+                {
+                    b.Property<string>("Id")
+                        .HasColumnType("text");
```
`text` fot the ID? not `varchar`?

jaredonline replied:
Is there a way to make the names of the columns be snake case and lowercase?

shmashing replied:
Yea we definitely could. EZ-PZ change, but will likely include with the next PR, if that's ok? I have a bunch of DB changes pending anyway

jaredonline commented on Destiny.ScrimTracker.Api/Requests/CreateGuardianRequest.cs:
```diff
@@ -0,0 +1,21 @@
+        {
+            return new Guardian()
+            {
+                Id = $"gua_{Guid.NewGuid():N}",
```
Can you explain this line to me?

jaredonline replied:
Can `gua` be in a constant somewhere?

shmashing replied:
Fasho! the '$' character denotes the beginning of a string interpolation. So we start our string out with the const 'gua_' and then we are telling it to create a new guid and append it to the string. the ':N' is a specific formatting condition for the string interpolation for GUIDs. It just means numbers and letters only, no dashes.

jaredonline commented on Destiny.ScrimTracker.Logic/Models/GuardianMatchResult.cs:
```diff
@@ -0,0 +1,21 @@
+        public Guid GuardianId { get; set; }
+        public Guid MatchId { get; set; }
+        public Guid MatchTeamId { get; set; }
+        public double Efficiency { get; set; }
```
Isn't `Efficiency` computed from `Kills` `Deaths` and `Assists`?

shmashing replied:
You thinks it may be redundant? mayhaps. but we will still want to store it, i think. Since the guardian_efficiences table tracks averages and not specific values

jaredonline commented on Destiny.ScrimTracker.Logic/Repositories/GuardianRepository.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            using (var context = _guardianContext)
+            {
+                var guardians =  context.Guardians.ToList();
```
Where is this method `Guardians` defined on `GuardianContext`?

shmashing replied:
It's not a method, it is a collection of objects. Entity Framework (our ORM) knows that when we are trying to access this collection, to read it in from the database first using LINQ (Language Integrated Query)

jaredonline commented on Destiny.ScrimTracker.Logic/Repositories/GuardianRepository.cs:
```diff
@@ -0,0 +1,100 @@
+        {
+            using (var context = _guardianContext)
+            {
+                var guardian = context.Guardians.FirstOrDefault(g => g.Id == guardianId);
```
This looks like its doing an in memory comparison and sort. Can we do it in SQL instead?

shmashing replied:
It is in SQL. The `FirstOrDefault` as well as `OrderBy` and `Where` extension methods all form SQL queries behind the hood that are then ran against the DB.

jaredonline replied:
What is the `g => g.Id == guardianId` bit? I assumed it was a lambda function, but its not?

shmashing replied:
No it still is. That lambda function is surfacing what field we want to query on

shmashing commented on Destiny.ScrimTracker.Api/Startup.cs:
```diff
@@ -25,6 +29,17 @@ public Startup(IConfiguration configuration)
+            // Register Services
+            services.AddTransient<IGuardianService, GuardianService>();
+            
             services.AddControllers();
```
@jaredonline RE: "Where does this [Controller Class] get instantiated?" This is where. We register all of our classes here as interfaces.

shmashing replied:
Controller classes are a special case where we don't have to register them individually. C# knows based on their location within the app
